### PR TITLE
[resourcedetectionprocessor] Add docker detector

### LINKED ISSUE #2773: System resource detection on containerized Collector
**Is your feature request related to a problem? Please describe.**

When running the containerized OpenTelemetry Collector, it is useful to have visibility into the host machine information, both for infrastructure monitoring and for enriching metadata on telemetry signals. The existing general-purpose system resource detector does not cover this use case since it only provide information about the container system. 

**Describe the solution you'd like and alternatives you've considered**

Either 

1. Have a new Docker detector that gathers this data or
2. add the functionality to the existing system resource detector, possibly through a flag.

Information that could be added is
- `host.name`
- `host.arch`
- `os.type`
- `container.runtime` (Docker)

**Additional context**

I will open a PR for this using the Docker Go SDK.

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

- Add Docker detector

**Link to tracking Issue:** Fixes #2773

**Testing:** Amended unit tests. Tested on Ubuntu 18.04 (you need to add `-v /var/run/docker.sock:/var/run/docker.sock`).

**Documentation:** Documented on README

### CODE REVIEW

**Comment by mx-psi** [comment]
@kbrockhoff can you take a look? Happy to try to improve test coverage after I have a review on the overall idea

**Comment by mx-psi** [comment]
@kbrockhoff Friendly ping!

Also @anuraaga @jrcamp I would be happy if you took a look :)

**Comment by bogdandrutu** [comment]
@jrcamp please review :)

File: processor/resourcedetectionprocessor/README.md [hunk 1]
```diff
@@ -17,6 +17,8 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+    If running the Collector Docker image, mount `/var/run/docker.sock` and set the `use_docker_sock` flag to `true` to retrieve the attributes from the host machine. Docker detection does not work on macOS.
```
Comment (jrcamp) [inline processor/resourcedetectionprocessor/README.md:20]: I kinda feel like it should be its own detector (you called it out as a possibility in the original issue). Was there any reason you decided not to go that route?

Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md:20]: > Was there any reason you decided not to go that route?

The main reasons were that 

1.  it felt like the Docker/bare system detectors should evolve together and provide the same attributes when technically feasible: we get the same attributes today and if e.g. someone wants to add the `host.arch` attribute when running the Collector binary it makes sense that they would also want it when running on Docker.
2. the system detector should work regardless of how the Collector is ran: from a user's POV it should make no difference if they are running the Collector from a bare binary or a containerized version. 

But as I wrote, I can definitely see it being its own detector, specially if we think they might end up diverging in some way (e.g. we want to add a `docker.version` attribute or some other attribute that only makes sense on Docker). We have to have a `use_docker_sock` flag anyway so it's not going to make the user experience much worse and it is more flexible

Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md]: @jrcamp I thought about this a bit more, and while I think the arguments above make sense, I think it's likely that these detector end up diverging so I separated it, PTAL

**Comment by tigrannajaryan** [comment]
@anuraaga @jrcamp please review.

File: processor/resourcedetectionprocessor/README.md [hunk 2]
```diff
@@ -17,6 +17,16 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+Use the Docker detector (see below) if running the Collector as a Docker container.
+
+* Docker metadata: Queries the Docker daemon to retrieve the following resource attributes from the host machine:
+
+    * host.name
```
Comment (anuraaga) [inline processor/resourcedetectionprocessor/README.md:24]: Can you describe the difference between retrieving these with normal means? I'm surprised we need to use the docker API to get simple looking attributes.

Maybe it's to retrieve the information about the host, not the docker container. If so, this seems wrong, I believe our semantic conventions intend these to be filled for the current process, which would be using the values from within the docker container, not the host machine. I'd expect us to define new conventions in the spec, something like `containerhost.host.name`, `containerhost.os.type` for things outside the process.

Sorry if I'm missing something

Comment (anuraaga) [reply processor/resourcedetectionprocessor/README.md:24]: Also I was expecting this PR to implement `container` attributes

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/container.md#container

Surprised to see `host.name` and `os.type` in the docker detector :)

Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md:24]: > Can you describe the difference between retrieving these with normal means? I'm surprised we need to use the docker API to get simple looking attributes.

Here is the output of the `hostname` command and the `Name` field from the Docker API info endpoint from an Ubuntu Docker image running on an Ubuntu VM:

```
❯ docker run --rm -v /var/run/docker.sock:/var/run/docker.sock -it ubuntu bash
root@3d639eae3af3:/# apt-get update && apt-get install -y curl jq
# ...
root@3d639eae3af3:/# hostname
3d639eae3af3
root@3d639eae3af3:/# curl --silent --unix-socket /var/run/docker.sock http://localhost/info | jq .Name
"ubuntu-vm"
```

`hostname` gives you the `os.Hostname` (which is a random string on Docker). The Docker API gives you the name of the host machine (on Linux and Windows).

Similarly, the `os.type` will give you the host OS type (so, e.g., `windows` even if your container is Linux-based).

> I'd expect us to define new conventions in the spec, something like containerhost.host.name, containerhost.os.type for things outside the process.

I can definitely do that, but I fail to see why someone would care about the Docker hostname: the whole point (at least to me) of running the Collector as a Docker image is to not have other stuff running in the same Docker image.

> Also I was expecting this PR to implement container attributes

👍 will do

Comment (anuraaga) [reply processor/resourcedetectionprocessor/README.md:24]: Ah, thanks for the explanation, that's the opposite behavior of what I was expecting for `os.type`, surprised native calls within the container don't return information about the container :O And populating something other than the random string seems good.

Just curious, do you know how common it is to have the docker API exposed? For example, will kubernetes users generally be able to use the detector?

**Review by anuraaga — APPROVED** [review]
I think we can add further `container` attributes in a future PR now I understand that this is important for correctly populating those existing ones

File: processor/resourcedetectionprocessor/README.md (hunk 2 above)
Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md:24]: > Just curious, do you know how common it is to have the docker API exposed? For example, will kubernetes users generally be able to use the detector?

I don't know about Kubernetes specifically, I would expect that people have to mount the Unix socket manually. We use this on our Datadog Agent for users running Docker by itself, but we have a different solution for Kubernetes.

**Comment by bogdandrutu** [comment]
Please rebase, codecov should not run :)

**Comment by bogdandrutu** [comment]
@jrcamp if you have any other comment please open an issue.

**Comment by mx-psi** [comment]
Sorry, I was out yesterday and didn't see this, happy to see there was no problem with the merge

**Pull request merged**